File: app/src/main/java/com/mrbimc/selinux/MainActivity.java
package com.mrbimc.selinux;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.AsyncTask;
import android.preference.PreferenceManager;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.CompoundButton;

import com.stericson.RootTools.RootTools;
import com.stericson.RootTools.execution.Command;

public class MainActivity extends AppCompatActivity {

    Button bEnforcing, bPermissive;
    CheckBox cNotify;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        bEnforcing = (Button) findViewById(R.id.button_enforcing);
        bPermissive = (Button) findViewById(R.id.button_permissive);
        cNotify = (CheckBox) findViewById(R.id.checkbox_notify);

        setupNotify();

        detectRootAccess();
    }

    private void setupNotify() {
        final SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
        int notify = sp.getInt("notify", 0);

        if (notify == 1) cNotify.setChecked(true);
        if (notify == 0) cNotify.setChecked(false);

        cNotify.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                if (isChecked) sp.edit().putInt("notify", 1).apply();
                else sp.edit().putInt("notify", 0).apply();
            }
        });
    }

    private void detectRootAccess() {
        new AsyncTask<Void, Void, Void>() {
            boolean isRoot = false;
            boolean isAccessGiven = false;

            @Override
            protected Void doInBackground(Void... params) {
                try {
                    isRoot = RootTools.isRootAvailable();
                    isAccessGiven = RootTools.isAccessGiven();
                } catch (Exception e) { e.printStackTrace(); }
                return null;
            }

            @Override
            protected void onPostExecute(Void aVoid) {
                super.onPostExecute(aVoid);
                if(isRoot && isAccessGiven) detectSELinuxMode();
                else {
                    bEnforcing.setEnabled(false);
                    bPermissive.setEnabled(false);
                    showDialog(getString(R.string.no_root_access), getString(R.string.no_root_access_summary));
                }

            }
        }.execute();
    }

    private void detectSELinuxMode() {

        Command command = new Command(0, "/system/bin/getenforce") {
            int status = 2;

            @Override
            protected void output(int id, String line) {
                super.output(id, line);
                status = line.contains(getString(R.string.selinux_enforcing)) ? 0 : line.contains(getString(R.string.selinux_permissive)) ? 1 : 2;
            }

            @Override
            public void commandCompleted(int id, int exitcode) {
                switch (status) {
                    case 0:
                        setEnforcing(null);
                        break;
                    case 1:
                        setPermissive(null);
                        break;
                    case 2:
                        setUndefined();
                        break;
                }
            }

            @Override
            public void commandOutput(int id, String line) {}
            @Override
            public void commandTerminated(int i, String s) {}
        };
        try { RootTools.getShell(true).add(command); } catch (Exception e) { e.printStackTrace(); }
    }

    public void setPermissive(View v){
        if(v != null) SELinuxBroadcastReceiver.permiss(this);
        bEnforcing.setEnabled(true);
        bPermissive.setEnabled(false);
    }

    public void setEnforcing(View v){
        if(v != null) SELinuxBroadcastReceiver.enforce(this);
        bEnforcing.setEnabled(false);
        bPermissive.setEnabled(true);
    }

    private void setUndefined(){
        bEnforcing.setEnabled(false);
        bPermissive.setEnabled(false);
        showDialog(getString(R.string.selinux_unreachable), getString(R.string.selinux_disabled));
    }

    private void showDialog(String title, String message){
        AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
        builder.setTitle(title)
                .setMessage(message)
                .setPositiveButton(getString(R.string.ok),null);
        builder.create();
        builder.setCancelable(false);
        builder.show();
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.action_about)
            startActivity(new Intent(this, CreditsActivity.class));
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/mrbimc/selinux/MainActivity.kt
package com.mrbimc.selinux

import android.content.Intent
import android.net.Uri
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.View
import com.yarolegovich.lovelydialog.LovelyProgressDialog
import android.preference.PreferenceManager
import android.view.Menu
import android.view.MenuItem
import android.widget.AdapterView
import android.widget.ArrayAdapter
import com.mrbimc.selinux.util.*
import com.yarolegovich.lovelydialog.LovelyStandardDialog
import kotlinx.android.synthetic.main.layout_command.*
import kotlinx.android.synthetic.main.layout_context.*
import kotlinx.android.synthetic.main.layout_notifications.*
import kotlinx.android.synthetic.main.layout_selinux_state.*
import kotlinx.android.synthetic.main.layout_autostart.*

class MainActivity : AppCompatActivity() {

    val sp by lazy { PreferenceManager.getDefaultSharedPreferences(applicationContext) }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        checkSELinuxState()
        checkCommand()
        checkContext()
        checkNotificationSettings()
        checkAutoStartSettings()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_about -> showInfoDialog()
        else -> false
    }

    fun setEnforcing(v: View) = setSELinuxState(SELinuxState.ENFORCING) {
        runOnUiThread {
            buttonEnforcing.isEnabled = false
            buttonPermissive.isEnabled = true
        }
    }

    fun setPermissive(v: View) = setSELinuxState(SELinuxState.PERMISSIVE) {
        runOnUiThread {
            buttonEnforcing.isEnabled = true
            buttonPermissive.isEnabled = false
        }
    }

    private fun checkContext() {
        val items = resources.getStringArray(R.array.selinux_contexts)
        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, items)


        val selectedItem = sp.getString(KEY_CONTEXT, items[0])
        val selectedItemPosition = adapter.getPosition(selectedItem)

        contextsSpinner.adapter = adapter
        contextsSpinner.setSelection(selectedItemPosition)

        contextsSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) { /*IGNORE*/ }

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                sp.edit().putString(KEY_CONTEXT, items[position]).apply()
            }
        }
    }

    private fun checkCommand() {
        val items = resources.getStringArray(R.array.selinux_commands)
        val adapter = ArrayAdapter(this, android.R.layout.simple_spinner_dropdown_item, items)

        val selectedItem = sp.getString(KEY_COMMAND, items[0])
        val selectedItemPosition = adapter.getPosition(selectedItem)

        commandsSpinner.adapter = adapter
        commandsSpinner.setSelection(selectedItemPosition)

        commandsSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(parent: AdapterView<*>?) { /*IGNORE*/ }

            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                sp.edit().putString(KEY_COMMAND, items[position]).apply()
            }
        }
    }

    private fun checkNotificationSettings() {
        notificationsSwitch.isChecked = sp.getBoolean(KEY_NOTIFICATIONS, true)

        notificationsSwitch.setOnCheckedChangeListener { _, isChecked ->
            sp.edit().putBoolean(KEY_NOTIFICATIONS, isChecked).apply()
        }
    }

    private fun checkAutoStartSettings() {
        autoStartSwitch.isChecked = sp.getBoolean(KEY_AUTOSTART, true)

        autoStartSwitch.setOnCheckedChangeListener { _, isChecked ->
            sp.edit().putBoolean(KEY_AUTOSTART, isChecked).apply()
        }
    }

    private fun checkSELinuxState() {

        val ad = LovelyProgressDialog(this)
                .setTopColorRes(R.color.colorAccent)
                .setTopTitle(getString(R.string.requesting_root))
                .setTopTitleColor(ContextCompat.getColor(this, android.R.color.white))
                .setIcon(R.drawable.security).show()

        getSELinuxState() {
            ad.cancel()

            when (it) {
                SELinuxState.UNKNOWN -> {
                    runOnUiThread {
                        buttonEnforcing.isEnabled = false
                        buttonPermissive.isEnabled = false
                    }
                }
                SELinuxState.PERMISSIVE -> {
                    runOnUiThread {
                        buttonEnforcing.isEnabled = true
                        buttonPermissive.isEnabled = false
                    }
                }
                SELinuxState.ENFORCING -> {
                    runOnUiThread {
                        buttonEnforcing.isEnabled = false
                        buttonPermissive.isEnabled = true
                    }
                }
            }
        }
    }

    private fun showInfoDialog(): Boolean {
        LovelyStandardDialog(this)
                .setTopColorRes(R.color.colorAccent)
                .setTopTitle(getString(R.string.about_this_app))
                .setTopTitleColor(ContextCompat.getColor(this, android.R.color.white))
                .setButtonsColorRes(R.color.colorAccent)
                .setIcon(R.drawable.information)
                .setMessage(R.string.info_dialog_message)
                .setNegativeButton(getString(R.string.button_close_dialog), null)
                .setPositiveButton(getString(R.string.button_open_github)) {
                    openGithub()
                }
                .show()

        return true
    }

    private fun openGithub() {
        startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://github.com/MrBIMC/SELinuxModeChanger")))
    }

}
